### PR TITLE
Feature   performance improvement

### LINKED ISSUE #57: Look into improving performance of producer
It would seem that the code is currently too slow to meet the desired throughput we need. Therefore some optimisations will need to be investigated.

### PR DESCRIPTION
**Pull request by Gala-Sortino**
Implements a more optimised queue system that should help provide more throughput with the same system resources. 

Also, it comments out the call needed to profile the sending threads created. 

Closes #57

### CODE REVIEW

File: src/SimpleProducer.py [hunk 1]
```diff
@@ -151,23 +154,46 @@ def data_pipe_producer(shared_data_queue, data_generator, max_queue_size, data_a
...
+            break
+
+def profile_senders(server_args,
+                           producer_counters,
+                           topic,
+                           shared_data_queue,
+                           avro_schema_keys,
+                           avro_schema_values,
+                           serializer,
+                           i):
```
Comment (gala-emile) [inline src/SimpleProducer.py]: Allign

File: src/SimpleProducer.py [hunk 2]
```diff
@@ -151,32 +154,54 @@ def data_pipe_producer(shared_data_queue, data_generator, max_queue_size, data_a
...
+    #                                serializer,
+    #                                i)) for i in range(numb_prod_procs)]
+    producer_procs = [Process(target=send,
+                              args=(server_args,
+                                    producer_counters,
+                                    topic,
+                                    shared_data_queue,
+                                    avro_schema_keys,
+                                    avro_schema_values,
+                                    serializer)) for _ in range(numb_prod_procs)]
```
[the remark this answers is not in the document]
Comment (Gala-Sortino) [reply src/SimpleProducer.py]: No easy way to shorten this.

File: src/DataConfiguration.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 from CSVReader import CSVReader
+from ParallelCSVReader import ParallelCSVReader
```
[the remark this answers is not in the document]
Comment (Gala-Sortino) [reply src/DataConfiguration.py]: This is fine.

File: src/SimpleProducer.py [hunk 4]
```diff
@@ -152,32 +168,73 @@ def data_pipe_producer(shared_data_queue, data_generator, max_queue_size, data_a
...
+
+    producer_procs = [Process(target=send,
+                              args=(server_args,
+                                    producer_counters,
+                                    topic,
+                                    shared_data_queue,
+                                    avro_schema_keys,
+                                    avro_schema_values,
+                                    serializer,
+                                    max_data_pipe_size)) for _ in range(numb_prod_procs)]
```
[the remark this answers is not in the document]
Comment (Gala-Sortino) [reply src/SimpleProducer.py]: Cannot resolve

File: src/SimpleProducer.py [hunk 5]
```diff
@@ -152,32 +168,73 @@ def data_pipe_producer(shared_data_queue, data_generator, max_queue_size, data_a
...
 
     shared_dict[topic] = manager.list()
-    shared_data_queue = manager.Queue()
-
+    #shared_data_queue = Queue(maxsize=max_data_pipe_size)
+    shared_data_queue = mp.Queue(maxsize=max_data_pipe_size)
     procs = []
 
-    data_gen_procs = [multiprocessing.Process(target=data_pipe_producer,
+    #data_gen_procs = [Process(target=profile_data_pipe_producer,
```
[the remark this answers is not in the document]
Comment (Gala-Sortino) [reply src/SimpleProducer.py]: This is expected

File: src/SimpleProducer.py [hunk 6]
```diff
@@ -215,6 +272,7 @@ def print_data_results(keys, dict_key):
         print("Mean: " + str(sum(item[key] for item in shared_dict[dict_key]) / length), end=' ')
         print("Max: " + str(max(item[key] for item in shared_dict[dict_key])), end=' ')
         print("Min: " + str(min(item[key] for item in shared_dict[dict_key])), end=' ')
+        print("Total: " + str(sum(item[key] for item in shared_dict[dict_key])), end=' ')
```
[the remark this answers is not in the document]
Comment (Gala-Sortino) [reply src/SimpleProducer.py]: No nice way to fix. Is ok as is.

**Pull request merged**